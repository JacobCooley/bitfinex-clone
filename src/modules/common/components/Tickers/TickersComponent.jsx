import React from 'react'
import { formatNumber } from 'functions/numbers'
import Button from 'component/Button/Button'
import Input from 'component/Input/Input'
import Loader from 'component/Loader'
import './Tickers.scss'

class TickersComponent extends React.Component {
	
	handleChangeValue = (e) => this.setState({ tickers: e.target.value })
	
	fetchTickers = () => {
		const tickersArray = this.state.tickers.split(",").map((item) => {
			const newItem = item.toUpperCase().trim()
			return newItem.substr(0, 1).toLowerCase() + newItem.substr(1, newItem.length)
		})
		this.props.fetchTickersJson(tickersArray)
	}
	
	render() {
		return (
			<>
				{this.props.showSpinner ? <Loader/> : null}
				<p>{`Enter Tickers you want to see. Separate with commas and prepend \'t\' or \'f\'`}</p>
				<Input label={`Tickers:`} name={`tickers`} onChange={this.handleChangeValue}/>
				<Button onClick={this.fetchTickers} text='Show Bitfinex tickers'/>
				<div className='tickers-list'>
					{this.props.tickersData.map((ticker, index) => {
							return (
								<ul key={Object.values(ticker[0])}>
									{
										ticker.map((value, i) => {
											return (
												<li key={Object.values(ticker[0]) + i}>{Object.keys(value)}: {isNaN(Object.values(value)) ? Object.values(value) : formatNumber(Object.values(value))}</li>
											)
										})
									}
								</ul>
							)
						}
					)}
				</div>
			</>
		)
	}
}

export default TickersComponent