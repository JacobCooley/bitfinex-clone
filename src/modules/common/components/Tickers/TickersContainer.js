import { connect } from 'react-redux'
import TickersComponent from './TickersComponent'
import { tickerOperations } from './duck/index'
import PropTypes from 'prop-types'

const mapStateToProps = (state) => {
	const { tickersData, showSpinner } = state.tickers
	return {
		tickersData,
		showSpinner
	}
}

const mapDispatchToProps = (dispatch) => {
	const fetchTickersJson = (tickers) => {
		dispatch(tickerOperations.fetchTickersJson(tickers))
	}
	return { fetchTickersJson }
}

const TickersContainer = connect(
	mapStateToProps,
	mapDispatchToProps
)(TickersComponent)

TickersContainer.propTypes = {
	tickersData: PropTypes.array,
	showSpinner: PropTypes.bool
}

TickersContainer.defaultProps = {
	tickersData: [],
	showSpinner: false
}

export default TickersContainer