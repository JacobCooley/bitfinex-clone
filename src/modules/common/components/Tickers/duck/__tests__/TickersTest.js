import TickersContainer from '../../TickersContainer'
import TickersComponent from '../../TickersComponent'
import { mapDispatchToProps } from '../../TickersContainer'
import React from 'react'
import Adapter from 'enzyme-adapter-react-16'
import { shallow, mount, configure } from 'enzyme'
import configureStore from 'redux-mock-store'
import renderer from 'react-test-renderer'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux'

configure({ adapter: new Adapter() })

describe('<TickersComponent /> TEST', () => {
	const state = {
		tickersData: [['tbtcusd', 12, 33, 123], ['tbtceth', 2, 32, .234, 423]]
	}
	it('+++ capturing Snapshot of <TickersComponent />', () => {
		const renderedValue = renderer.create(<TickersComponent tickersData={[[0, 12, 33, 123]]}/>).toJSON()
		expect(renderedValue).toMatchSnapshot()
	})
	
	it('+++ capturing Snapshot of <TickersComponent /> with loading', () => {
		const renderedValue = renderer.create(<TickersComponent tickersData={[[0, 12, 33, 123]]}
																showSpinner={true}/>).toJSON()
		expect(renderedValue).toMatchSnapshot()
	})
	
	it('should call handleChangeValue prop', () => {
		const wrapper = mount(<TickersComponent {...state} />)
		wrapper.find('input').simulate('change', { target: { value: 'Your new Value' } })
		expect(wrapper.state('tickers')).toBe('Your new Value')
	})
})


describe('<TickersContainer /> mount copy', () => {
	const initialState = {
		tickers: {
			tickersData: [],
			showSpinner: false
		}
	}
	const middlewares = [thunk]
	const mockStore = configureStore(middlewares)
	let store, wrapper
	
	beforeEach(() => {
		store = mockStore(initialState)
		wrapper = mount(<Provider store={store}><TickersContainer/></Provider>)
	})
	
	it('render the connected component', () => {
		expect(wrapper.find(TickersContainer).length).toEqual(1)
	})
	
	it('check Prop matches with initialState', () => {
		expect(wrapper.find(TickersComponent).prop('tickersData')).toEqual(initialState.tickers.tickersData)
		expect(wrapper.find(TickersComponent).prop('showSpinner')).toEqual(initialState.tickers.showSpinner)
	})
})