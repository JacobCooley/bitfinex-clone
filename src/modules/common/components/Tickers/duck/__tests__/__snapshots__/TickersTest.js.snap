// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TickersComponent /> TEST +++ capturing Snapshot of <TickersComponent /> 1`] = `
Array [
  <p>
    Enter Tickers you want to see. Separate with commas and prepend 't' or 'f'
  </p>,
  <label>
    Tickers:
    <input
      name="tickers"
      onChange={[Function]}
      type="text"
    />
  </label>,
  <button
    onClick={[Function]}
  >
    Show Bitfinex tickers
  </button>,
  <div
    className="tickers-list"
  >
    <ul>
      <li>
        : 
        NaN
      </li>
      <li>
        : 
        NaN
      </li>
      <li>
        : 
        NaN
      </li>
      <li>
        : 
        NaN
      </li>
    </ul>
  </div>,
]
`;

exports[`<TickersComponent /> TEST +++ capturing Snapshot of <TickersComponent /> with loading 1`] = `
Array [
  <p>
    Enter Tickers you want to see. Separate with commas and prepend 't' or 'f'
  </p>,
  <label>
    Tickers:
    <input
      name="tickers"
      onChange={[Function]}
      type="text"
    />
  </label>,
  <button
    onClick={[Function]}
  >
    Show Bitfinex tickers
  </button>,
  <div
    className="tickers-list"
  >
    <ul>
      <li>
        : 
        NaN
      </li>
      <li>
        : 
        NaN
      </li>
      <li>
        : 
        NaN
      </li>
      <li>
        : 
        NaN
      </li>
    </ul>
  </div>,
]
`;
